File: player/src/sprite.cpp
#include <frames.h>
#include <sprite.h>
#include <loop.h>

Sprite::Sprite(Frames *_frames, Loop *_loops) :
    frames(_frames), loops(_loops) {
}

Loop* Sprite::loopAtIndex(uint8_t index) {
  Loop *loop = &loops[index];
  return loop;
}

void Sprite::draw(uint8_t x, uint8_t y, uint8_t frame) {
  frames->draw(x, y, frame);
}

File: export/lib/model/project/entity/sprite.js
const Entity = require('../entity');
const array = require('./array');

const filtered = (sprite, type) => array(sprite.project.sortBy(sprite.models.filter(model => model.type === type), 'json.index', 'asc'));

class Sprite extends Entity {

  bind() {
    super.bind();
    this.index = this.parent.sprites.indexOf(this);
    this.size = this.json.size;
    this.frames = filtered(this, 'sprite/frame');
    this.loops = filtered(this, 'sprite/loop');
  }

  toPlusMaskString() {
    let { size, frames } = this;
    let data = frames.map(frame => frame.bytes.toPlusMaskString())
    return [
      `${size.width}, ${size.height},`,
      data.join(',\n')
    ].join('\n');
  }

}

module.exports = Sprite;
